File: Leetcode/Partition Array For Maximum Sum/Partition Array For Maximum Sum.cpp

/*make one array dp of length same as A, and fill this with 0
for i in range 0 to length of A - 1
dp[i] = A[i] + dp[i - 1] if i – 1 >= 0 otherwise 0
temp := A[i]
for j in range 1 to k – 1
if i – j >= 0
index := i – j
temp := max of temp and A[i - j]
if index – 1 >= 0, then
dp[i] := max of dp[i] and (temp*(i – index + 1) + dp[index - 1])
otherwise dp[i] := max of dp[i] and 0
return last element of dp*/
class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& A, int K) {
int n = A.size();
    vector<int> dp(n);
    for (int i = 0; i < n; ++i) {
        int curMax = 0;
        for (int j = 1; j <= K && i - j + 1 >= 0; ++j) {
            curMax = max(curMax, A[i - j + 1]);
            dp[i] = max(dp[i], (i >= j ? dp[i - j] : 0) + curMax * j);
        }
    }
    return dp.back();
    }
};

File: Leetcode/Partition Array For Maximum Sum/Partition Array for Maximum Sum.java
class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        //Finding the size of given array
        int size = arr.length;
        //Dp array will contain the answer for number of elements at each index i
        int[] dp = new int[size + 1];
        //Looping through the array
        for (int i = 1; i <= size; ++i) {
            //temp will store the temporary answer
            int temp= 0; 
            //loopAns will store the answer for the loop
            int loopAns = 0;
            //We will loop for each value of j from1 to k
            //Then we will store the temporary maximum and also update our loop answer
            for (int j = 1; j <= k && i - j >= 0; j++) {
                temp = Math.max(temp, arr[i - j]);
                loopAns = Math.max(loopAns, dp[i - j] + temp * j);
            }
            //Storing the value
            dp[i] = loopAns;
        }
        return dp[size];
    }
}
